src/ir/auth_logic/ast.h

```diff
@@ -0,0 +1,176 @@
+#include <variant>
+#include <vector>
+
+namespace raksha::ir::auth_logic {
+
+class Principal {
+ public:
+  Principal(std::string name) : name_(std::move(name)) {}
+
+ private:
+  std::string name_;
+};
+
+// Used to represent whether a predicate is negated or not
+enum Sign { kNegated, kPositive };
+
+class Predicate {
+ public:
+  Predicate(std::string name, std::vector<std::string> args, Sign sign)
+      : name_(std::move(name)),
+        args_(std::move(args)),
+        // TODO move surpresses an unused private field error about sign_
+        // for now, get rid of this eventually
+        sign_(std::move(sign)) {}
+
+ private:
+  std::string name_;
+  std::vector<std::string> args_;
+  Sign sign_;
+};
+
+// This represents an expression that applies either a predicate or a
+// "canActAs" to a principal, and it is a sub-term of FlatFact
+//    - pred_name(arg_1, ..., arg_n)           (a predicate)
+//    - principal_1 canActAs principal_2       (a canActAs)
+// To encapsulate which form this takes, it:
+//    - uses factory methods to make either form
+//    - uses a private std::variant member that takes either of these two forms
+//    - uses std::visit to do operations on the actual state
+class VerbPhrase {
+ public:
+  explicit VerbPhrase(Predicate pred) : internal_form_(std::move(pred)) {}
+  explicit VerbPhrase(Principal right_prin)
+      : internal_form_(std::move(right_prin)) {}
+
+ private:
+  // The principal variant of the internal form means this represents
+  // "canActAs principal".
+  std::variant<Predicate, Principal> internal_form_;
+};
+
+// This the AST Node for FlatFact which takes on one of two forms:
+//   - <principal> <verb_phrase>  (* a principal applied to a verbphrase *)
+//   - <predicate>                (* just a predicate *)
+// A similar variant-based pattern is used for this as for VerbPhrase. Here an
+// inner struct is used for the first form (principal applied to verb_phrase).
+class FlatFact {
+ public:
+  explicit FlatFact(Principal prin, VerbPhrase verb_phrase)
+      : internal_form_(
+            VerbPhraseFlatFact(std::move(prin), std::move(verb_phrase))) {}
+  explicit FlatFact(Predicate pred) : internal_form_(std::move(pred)) {}
+
+ private:
+  struct VerbPhraseFlatFact {
+    Principal prin;
+    VerbPhrase verb_phrase;
+    VerbPhraseFlatFact(Principal prin, VerbPhrase verb_phrase)
+        : prin(std::move(prin)), verb_phrase(std::move(verb_phrase)) {}
+  };
+  std::variant<VerbPhraseFlatFact, Predicate> internal_form_;
+};
+
+// This is the AST node for facts which have two forms:
+//  - <flat_fact>
+//  - <principal> canSay <flat_fact>
+class Fact {
+ public:
+  explicit Fact(FlatFact flat_fact) : internal_form_(std::move(flat_fact)) {}
+  explicit Fact(Principal prin, Fact* fact)
+      : internal_form_(CanSayFact(std::move(prin), std::move(fact))) {}
+
+ private:
+  struct CanSayFact {
+    Principal prin;
+    std::unique_ptr<Fact> fact;
+    CanSayFact(Principal prin, Fact* fact)
```

Comment:
This should be `std::unique_ptr<Fact>` and not `Fact*`.

Reply:
I made these shared pointers so that copies can be made for accessors, which seems better than having accessors take ownership of the memory.